lib/ace/autocomplete.js

```diff
@@ -479,7 +479,7 @@ var FilteredList = function(array, filterText) {
         var upper = needle.toUpperCase();
         var lower = needle.toLowerCase();
         loop: for (var i = 0, item; item = items[i]; i++) {
-            var caption = item.value || item.caption || item.snippet;
+            var caption = item.snippet || item.caption || item.value;
```

Comment:
this breaks the completer for snippets, should we use caption || value || name to match https://github.com/ajaxorg/ace/blob/master/lib/ace/autocomplete/popup.js#L181 ?

Reply:
can you give an example of when it breaks? and what `item.name` is? 

the change was the make these two consistent:
https://github.com/ajaxorg/ace/blob/2b263a4a23050a3f61303308fbd7e66d2c658f5e/lib/ace/autocomplete.js#L469
https://github.com/ajaxorg/ace/blob/2b263a4a23050a3f61303308fbd7e66d2c658f5e/lib/ace/autocomplete.js#L482

Reply:
this two are intentionally inconsistent to allow different snippets with same caption.
But, if we want to fix the issue of `users were searching the value they weren't being shown` we need to make this consistent with the order of how things are shown in the popup.

If snippet is tested first, then captions of snippets will be shown, but contents used in the filter.

> and what item.name is?

seems like it is for backwards compatibility, so we can ignore it here.

Reply:
ah i think i understand roughly.

the main problem is that we highlight matches in the popup based on substring in the **value**, whereas we display the **caption** so its possible we highlight the wrong substring if the two differ.

Reply:
@nightwing im struggling to understand what you suggest we do from here actually. the snippets autocompleter checks `value || name`, so are you suggesting we just have `snippet || caption || value || name` here?

or are you suggesting we change the snippets autocompletion code?

Reply:
I am suggesting to change this line to `item.value || item.caption || item.snippet;` to match the way items are displayed. 
Would this fix the issue you are seeing?

Reply:
this line (that we are commenting on now) was value/caption/snippet before. that caused the inconsistency. we need to be searching the caption rather than the value i think, so what you see is what you're searching (this means the bold tokens line up correctly too).

so is it `popup.js` that needs changing to match? im not too sure but if we change this line back to `value` it would re-introduce the bug.

Reply:
I agree that  caption needs to be before value, i am saying snippet needs to be after everything else, because it is not displayed if there is caption or value.

Reply:
ah! so `caption || value || snippet` could sort that? but we'd have to change both lines i mentioned above to match

Reply:
the  line https://github.com/ajaxorg/ace/blob/2b263a4a23050a3f61303308fbd7e66d2c658f5e/lib/ace/autocomplete.js#L469 does not need to match this line, because it is for filtering duplicate entries, and we want to allow entries with same caption but different value/snippet

Reply:
ah that is true. so we can probably just change the other to include snippet on the end?

though it means you will see the value instead of the snippet, is that correct?

Reply:
we basically need to ensure that this line (482) is in the same order as what we display.

so what you search is what you see. so if i move it to `caption || value || snippet`, we should also make sure thats the order we use when displaying it. but im unsure where that logic lives

Reply:
you already see caption or value instead of snippet, since this is controlled in popup.js and snippet is usually multiline anyway

Reply:
great, ill change this to `caption || value || snippet` then and we are sorted i think.

thanks a lot for explaining that and giving some of your time. was a little confusing to figure out